File: backend/src/entropia/application/queries/trade_log.py
"""Trade Log read-side queries (Stage 3d, doc 05 §8, §10).

Two read surfaces:

* ``get_import_report`` — the durable import status + accepted/skipped counts +
  skipped-row evidence for a job (doc 05 §6 Import Report).
* ``get_trade_log`` — the native work object header + current §10.2 revision payload.

Both are authentication-gated and re-authorize ownership server-side; the report is
scoped to the source asset's owner, the detail to the work object's owner.
"""

from __future__ import annotations

from typing import Any

from sqlalchemy.ext.asyncio import AsyncSession

from entropia.domain.identity import Actor
from entropia.domain.identity.policy import ensure_can_view, require_authenticated
from entropia.infrastructure.postgres.models import Job
from entropia.infrastructure.postgres.repositories import mainboard as mb_repo
from entropia.infrastructure.postgres.repositories import trade_log as tl_repo
from entropia.shared.errors import (
    ImportJobNotFoundError,
    WorkObjectNotFoundError,
)

_MAX_SKIPPED_ROWS = 200


async def get_import_report(session: AsyncSession, actor: Actor, *, job_id: str) -> dict[str, Any]:
    """Return the durable import status + evidence for a job (doc 05 §6, §8).

    While the worker is still running there is no record batch yet — the job status
    is returned. Once complete, the batch's counts + skipped-row report + validation
    summary are surfaced (skipped rows are capped so a huge file cannot blow up the
    response, doc 05 §3.2).
    """
    require_authenticated(actor)
    job = await session.get(Job, job_id)
    if job is None:
        raise ImportJobNotFoundError(f"Import job '{job_id}' not found.")
    ensure_can_view(actor, owner_principal_id=job.actor_principal_id, visibility="private")

    batch = await tl_repo.get_record_batch_for_job(session, job_id)
    if batch is None:
        return {
            "job_id": job_id,
            "status": str(job.status),
            "record_batch_revision_id": None,
            "accepted_count": 0,
            "skipped_count": 0,
            "skipped_rows": [],
            "validation_summary": None,
        }
    return {
        "job_id": job_id,
        "status": str(batch.status),
        "record_batch_revision_id": batch.record_batch_id,
        "source_asset_id": batch.source_asset_id,
        "instrument_id": batch.instrument_id,
        "accepted_count": batch.accepted_count,
        "skipped_count": batch.skipped_count,
        "skipped_rows": (batch.skipped_rows or [])[:_MAX_SKIPPED_ROWS],
        "validation_summary": batch.validation_summary,
        "content_hash": batch.content_hash,
    }


async def get_trade_log(session: AsyncSession, actor: Actor, *, root_id: str) -> dict[str, Any]:
    """Return the Trade Log work object header + current revision (doc 05 §10.1)."""
    require_authenticated(actor)
    root = await mb_repo.get_work_object_root(session, root_id)
    detail = await mb_repo.get_work_object_detail(session, root_id)
    if root is None or detail is None or str(detail.object_kind) != "trade_log":
        raise WorkObjectNotFoundError(f"Trade Log '{root_id}' not found.")
    ensure_can_view(actor, owner_principal_id=root.owner_principal_id, visibility="private")

    revision = None
    if root.current_revision_id is not None:
        revision = await mb_repo.get_work_object_revision(session, root.current_revision_id)

    return {
        "root_id": root.entity_id,
        "object_kind": str(detail.object_kind),
        "owner_principal_id": root.owner_principal_id,
        "lifecycle_state": root.lifecycle_state,
        "deletion_state": str(root.deletion_state),
        "current_revision_id": root.current_revision_id,
        "row_version": root.row_version,
        "current_revision": _revision_projection(revision) if revision is not None else None,
    }


def _revision_projection(revision: Any) -> dict[str, Any]:
    return {
        "revision_id": revision.revision_id,
        "revision_no": revision.revision_no,
        "payload": revision.payload,
        "source_provenance": revision.source_provenance,
        "available_time": (
            revision.available_time.isoformat() if revision.available_time is not None else None
        ),
        "content_hash": revision.content_hash,
    }


__all__ = ["get_import_report", "get_trade_log"]

File: frontend/src/lib/tradeLog.ts
// Trade Log data access (doc 05 §8–§10). Typed read models + mutations over
// routes/trade_log.py — the near-symmetric twin of lib/tradingSignal.ts with
// two deliberate divergences mirrored VERBATIM from the backend:
//  1. The import produces a canonical_trade_record_batch — the report and the
//     §10.2 import_binding carry record_batch_revision_id (NOT
//     normalized_event_revision_id).
//  2. Trade Log is HISTORICAL data (doc 05 §10.4): the work-object revision
//     always carries available_time=null — there is no per-event
//     anti-lookahead availability contract.
// EMPIRICAL route map (signatures read): identical shape to the Trading
// Signal router — 4 POSTs all read a fresh Idempotency-Key; the ONLY OCC
// token is BODY-form expected_head_revision_id STR on /revisions (server-side
// optional — this client ALWAYS sends the rendered head); create/upload/
// import carry no OCC (no head to race). Pin ("Use This Revision") and delete
// REUSE the Mainboard router (CR-01, TL-01: a Trade Log is a work object, not
// a package). The import report is keyed under ["jobs"] (job.updated SSE
// sweep + terminal-stop poll fallback, INF-11); the work-object keys
// ["trade-logs"] have no dedicated SSE event → resource.changed sweeps them.

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api, apiRequest } from "./apiClient";
import { uploadFile } from "./upload";
import type {
  RequestImportResult,
  UploadSourceAssetResult,
  WorkObjectRevisionView,
} from "./tradingSignal";
import { TERMINAL_IMPORT_STATUSES } from "./tradingSignal";

// Shared column-mapping helpers (doc 05 §5.2) — re-exported so the Trade Log page
// imports them from its own lib, mirroring the twin's surface.
export { mappingHashFromSummary, parseColumnMapping } from "./tradingSignal";

// ---------------------------------------------------------------------------
// Wire types (mirror application/{commands,queries}/trade_log.py returns)
// ---------------------------------------------------------------------------

// GET /trade-logs/imports/{job_id} (queries get_import_report). The twin diff:
// the produced evidence is a record batch, not a normalized event revision.
export interface TradeLogImportReport {
  job_id: string;
  status: string;
  record_batch_revision_id: string | null;
  source_asset_id?: string;
  instrument_id?: string;
  accepted_count: number;
  skipped_count: number;
  skipped_rows: Array<Record<string, unknown>>;
  validation_summary: Record<string, unknown> | null;
  content_hash?: string;
}

// GET /trade-logs/{root_id} (queries get_trade_log). current_revision reuses
// the shared work-object projection; available_time is always null here.
export interface TradeLogDetail {
  root_id: string;
  object_kind: string;
  owner_principal_id: string | null;
  lifecycle_state: string;
  deletion_state: string;
  current_revision_id: string | null;
  row_version: number;
  current_revision: WorkObjectRevisionView | null;
}

// POST /trade-logs return (commands create_trade_log_and_attach).
export interface CreateTradeLogResult {
  root_id: string;
  revision_id: string;
  revision_no: number;
  object_kind: string;
  config_hash: string;
  row_version: number;
  attached: boolean;
  ready_state: string;
  workspace_id?: string;
  item_id?: string;
  composition_hash?: string | null;
}

// POST /trade-logs/{root_id}/revisions return (commands
// create_trade_log_revision). auto_repinned is always false (Rule 10).
export interface CreateTradeLogRevisionResult {
  root_id: string;
  revision_id: string;
  revision_no: number;
  config_hash: string;
  row_version: number;
  auto_repinned: boolean;
}

// ---------------------------------------------------------------------------
// Taxonomy mirrors (hydration only — the server re-validates every dispatch)
// ---------------------------------------------------------------------------

// commands/trade_log.py _ALLOWED_EXTENSIONS.
export const TRADE_LOG_SOURCE_EXTENSIONS = [".txt", ".csv"] as const;

const IMPORT_POLL_INTERVAL_MS = 5000;

// §10.2 payload skeleton pre-seeded with the import binding — a hydration-only
// convenience for the JSON editor. The twin diffs vs the Trading Signal
// template: time_model (not time_policy + event_model), content_profile data
// quality, ledger price source, and the record-batch import binding.
export function buildTradeLogPayloadTemplate(input: {
  sourceAssetId: string;
  recordBatchRevisionId: string;
  instrumentId: string;
  sourceTimezone: string;
  mappingRevisionId?: string;
}): Record<string, unknown> {
  const importBinding: Record<string, unknown> = {
    source_asset_id: input.sourceAssetId,
    record_batch_revision_id: input.recordBatchRevisionId,
  };
  if (input.mappingRevisionId !== undefined && input.mappingRevisionId !== "") {
    importBinding.mapping_revision_id = input.mappingRevisionId;
  }
  return {
    kind: "trade_log",
    identity: { display_name: "" },
    source: { provider_name: "", source_kind: "file" },
    instrument_scope: {
      instrument_id: input.instrumentId,
      display_symbol: input.instrumentId,
    },
    time_model: {
      resolution_kind: "event_based",
      source_timezone: input.sourceTimezone,
      normalization_timezone: "UTC",
    },
    data_quality: { content_profile: "entry_exit_records_only" },
    price_policy: { source: "trade_log_entry_exit_price" },
    ohlcv_policy: { use_mode: "ignore" },
    import_binding: importBinding,
  };
}

// ---------------------------------------------------------------------------
// Query hooks
// ---------------------------------------------------------------------------

export function useTradeLog(rootId: string | null) {
  return useQuery({
    queryKey: ["trade-logs", "root", rootId],
    queryFn: () => api.get<TradeLogDetail>(`/trade-logs/${encodeURIComponent(rootId ?? "")}`),
    enabled: rootId !== null,
  });
}

// ["jobs"]-keyed like the Trading Signal report: job.updated sweeps it live;
// the poll fallback stops on a terminal status (INF-11).
export function useTradeLogImportReport(jobId: string | null) {
  return useQuery({
    queryKey: ["jobs", "trade-log-import", jobId],
    queryFn: () =>
      api.get<TradeLogImportReport>(`/trade-logs/imports/${encodeURIComponent(jobId ?? "")}`),
    enabled: jobId !== null,
    refetchInterval: (query) => {
      const status = query.state.data?.status;
      return status !== undefined && TERMINAL_IMPORT_STATUSES.has(status)
        ? false
        : IMPORT_POLL_INTERVAL_MS;
    },
  });
}

// ---------------------------------------------------------------------------
// Mutations
// ---------------------------------------------------------------------------

// Upload the immutable raw TXT/CSV source asset (content-addressed dedup,
// TL-15). Return shape is shared with the Trading Signal twin.
export function useUploadTradeLogSource() {
  const queryClient = useQueryClient();
  return useMutation({
    // F-03: real native file transfer (multipart) — the browser sends the
    // selected TXT/CSV trade-record file itself; the server derives content type
    // + digest and validates size/encoding/schema. No pasted-textarea content.
    mutationFn: (input: { file: File }) =>
      uploadFile<UploadSourceAssetResult>("/trade-logs/source-assets", input.file, {
        idempotencyKey: crypto.randomUUID(),
      }).promise,
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

// 202 durable import job on the data queue (TL-14/TL-15 idempotent; the job id
// is the durable handle — keep it in the URL, CR-09).
export function useRequestTradeLogImport() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: {
      sourceAssetId: string;
      instrumentId: string;
      sourceTimezone: string;
      importMapping?: Record<string, string> | null;
    }) =>
      apiRequest<RequestImportResult>("/trade-logs/imports", {
        method: "POST",
        body: {
          source_asset_id: input.sourceAssetId,
          instrument_id: input.instrumentId,
          source_timezone: input.sourceTimezone,
          ...(input.importMapping && Object.keys(input.importMapping).length > 0
            ? { import_mapping: input.importMapping }
            : {}),
        },
        headers: { "Idempotency-Key": crypto.randomUUID() },
      }),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["jobs"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

// Save (& Add): create the native work object + revision 1 (Save != Ready
// PASS != Run, Implementation Rule 15). Attached → composition_hash moved →
// ["mainboard"] + ["readiness"] join the sweep.
export function useCreateTradeLog() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: { payload: Record<string, unknown>; attach: boolean }) =>
      apiRequest<CreateTradeLogResult>("/trade-logs", {
        method: "POST",
        body: { payload: input.payload, attach: input.attach },
        headers: { "Idempotency-Key": crypto.randomUUID() },
      }),
    onSuccess: (result) => {
      void queryClient.invalidateQueries({ queryKey: ["trade-logs"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
      if (result.attached) {
        void queryClient.invalidateQueries({ queryKey: ["mainboard"] });
        void queryClient.invalidateQueries({ queryKey: ["readiness"] });
      }
    },
  });
}

// Append immutable revision N+1 (Rule 10: NEVER auto-repins). The rendered
// head always travels as the BODY-form OCC token.
export function useCreateTradeLogRevision() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: {
      rootId: string;
      payload: Record<string, unknown>;
      expectedHeadRevisionId: string;
    }) =>
      apiRequest<CreateTradeLogRevisionResult>(
        `/trade-logs/${encodeURIComponent(input.rootId)}/revisions`,
        {
          method: "POST",
          body: {
            payload: input.payload,
            expected_head_revision_id: input.expectedHeadRevisionId,
          },
          headers: { "Idempotency-Key": crypto.randomUUID() },
        },
      ),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["trade-logs"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

// Export (S6, doc 05 §8 "Export As Package", §11, §13.2) — the immutable
// source-mapping/provenance MANIFEST for a Trade Log revision. Twin of
// useExportTradingSignal / the R2c package export: synchronous, owner/Admin-gated
// server-side, read-only provenance (writes only a trade_log.exported audit; the
// work object is unchanged) → invalidates only ["audit"]. A fresh Idempotency-Key
// makes repeated clicks return the same manifest_hash; NO OCC token (revisionId
// omitted → the server exports the pinned head).
export interface ExportTradeLogResult {
  root_id: string;
  revision_id: string;
  manifest_hash: string;
  manifest: Record<string, unknown>;
}

export function useExportTradeLog() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: { rootId: string; revisionId?: string }) =>
      apiRequest<ExportTradeLogResult>(
        `/trade-logs/${encodeURIComponent(input.rootId)}/export`,
        {
          method: "POST",
          headers: { "Idempotency-Key": crypto.randomUUID() },
          body:
            input.revisionId !== undefined
              ? { revision_id: input.revisionId }
              : {},
        },
      ),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}
